apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs

```diff
@@ -115,20 +140,236 @@ private PooledSession(SessionPool.ISessionPool pool, SessionName sessionName, By
+            if (SingleUseTransaction)
+            {
+                transactionSelectorSetter(new TransactionSelector { SingleUse = TransactionOptions });
+                return await commandAsync().ConfigureAwait(false);
+            }
+
+            // If we already have a transaction ID, we set the selector to that and execute the command.
+            // TransactionId is accessed and modified via Interlock.CompareExchange so these two are atomic operations.
+            // But also, if TransactionId is about to be modified right after this check, that's not a problem, because next
+            // we'll be awaiting on the task that does the modifying.
+            if (TransactionId is ByteString transactionId)
+            {
+                transactionSelectorSetter(new TransactionSelector { Id = transactionId });
+                return await commandAsync().ConfigureAwait(false);
+            }
+
+            // We now know that we don't have a transaction ID but we need one to execute
+            // the command we have been given.
+            // There are 
+            // We now need to check if we are already creating a transaction or not.
+            // If we are, we just get ready to wait for it.
+            // If we are not, but we need to, we start and save the task that does so,
+            // and get ready to wait for it.
+
+            // This is the function that will wait for the transaction task to be done.
+            // We need to initialize a function within the lock, so we can execute
+            // async code, outside the lock.
+            // We initialize it with just command, in case no transaction is being created
+            // and the caller does not require one. This is the case for commits and rollbacks
+            // executed when no transaction has been created before.
+            // Commits and rollbacks will know how to handle transaction absence.
+            Func<Task<TResponse>> commandMaybeWithTransactionAsync = commandAsync;
+
+            lock (_transactionCreationTaskLock)
+            {
+                // We are not creating a transaction. We might need to do so.
+                if (_transactionCreationTask is null)
+                {
+                    // We need to create a transaction.
+                    if (!skipTransactionCreation)
+                    {
+                        // The calling command does not support inilining
+                        // or the transaction mode Partitioned DML, which cannot be inline.
+                        // Either way we need to create a transaction explicitly.
+                        if (inlinedTransactionExtractor is null || TransactionOptions.ModeCase == ModeOneofCase.PartitionedDml)
+                        {
+                            _transactionCreationTask = Task.Run(() => SetExplicitTransactionAsync(cancellationToken), cancellationToken);
```

Comment:
I'm nervous about calling Task.Run in a lock. There are always weird cases where it runs synchronously. Not sure of an alternative though.